dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+        } else {
+            --inst->length;
+        }
+    } else {
+        buffer_reserve(inst, inst->alloc_by + length);
+    }
+}
+
+// Pointer to first item in buffer
+void *buffer_front(struct buffer *inst)
+{
+    if (inst) return inst->buffer[0];
+}
+
+// Pointer to last item in buffer
+void *buffer_back(struct buffer *inst)
+{
+    if (inst) return buffer_at(inst, buffer_size(inst) - 1);
+}
+
+// Take last item from buffer
+//  * return false if buffer is empty
+//  * copy item to "out" if not null
+//  * call destructor for item being removed
+//  * reduce length by 1
+//  * return true
+bool buffer_pop(struct buffer *inst, void *out)
+{
+    if (buffer_empty(inst)){
+        return false;
+    } else if (!buffer_back(inst)) {
+        return false;
+    } else {
+        memcpy(out, buffer_back(inst), inst->item_size);
+        --inst->length;
+        return true;
+    }
+}
+
+// Return true if buffer is empty (length == 0)
+bool buffer_empty(struct buffer *inst){
+    if (inst){
+       return (buffer_size(inst) == 0); 
+    } else {
+        exit(0);
+    }
+
+}
+
+// Number of items in the buffer (length)
+size_t buffer_size(struct buffer *inst)
+{
+    if (inst) {
+        return inst->length;
+    } else {
+        exit(0);
+    }
+}
+
+// Item size of the elements in the buffer(item_size)
+size_t item_size(struct buffer *inst)
+{
+    if (inst) {
+        return inst->item_size;
+    } else {
+        exit(0);
+    }
+}
+
+// Pointer to nth item in buffer
+void *buffer_at(struct buffer *inst, size_t index)
+{
+    if (index >= inst->length) {
+        return NULL;
+    }
+    return (inst->buffer[index]);
+}
+
+void destruct_size_t_item(void *item){
+    free(item);
+}
+
+
+
+// TESTING
+#if defined TEST_DYNAMIC_ARRAY
+
+struct buffer init()
+{
+    struct buffer inst;
+    destr destruct_func = &destruct_size_t_item;
+    buffer_init(&inst, sizeof(int*), 5, 2, destruct_func);
+    return inst;
+}
+
+void test_init()
+{
+    struct buffer inst = init();
+    assert(buffer_empty(&inst));
+    assert(buffer_size(&inst) == 0);
+    assert(item_size(&inst) == 8);
+    assert(buffer_at(&inst, 89) == NULL);
+
+    buffer_destroy(&inst);
+}
+
+void test_add_elem()
+{   
+    struct buffer inst = init();
+    int *p = malloc(sizeof (int*));
+    *p = 64;
+    buffer_push_back(&inst, p);
+
+    assert(!buffer_empty(&inst));
+    assert(buffer_size(&inst) == 1);
+    assert(*(int*)buffer_at(&inst, 0) == 64);
+    assert(buffer_front(&inst) == buffer_back(&inst));
+    free(p);
+    buffer_destroy(&inst);
+}
+
+void test_pop_elem()
+{
+    struct buffer inst = init();
+    int *p = malloc(sizeof (int*));
+    *p = 64;
+    buffer_push_back(&inst, p);
+
+    void *out = malloc(sizeof (int*));
+    assert(buffer_pop(&inst, out));
+
+    assert(buffer_empty(&inst));
+    assert(buffer_size(&inst) == 0);
+    assert(*(int*)out == 64);
+
+    free(out);
+    free(p);
+    buffer_destroy(&inst);
+}
+
+void test_resize()
```

Comment:
Functions which are not intended to be used from outside this C file should be marked as `static`, to prevent them from being exported.  This in turn prevents name collisions during linking.